ehrql/backends/tpp.py

```diff
@@ -22,6 +25,62 @@ class TPPBackend(BaseBackend):
         ehrql.tables.beta.smoketest,
     ]
 
+    include_t1oo = False
+
+    def modify_dsn(self, dsn):
+        """
+        Removes the `opensafely_include_t1oo` parameter if present and uses it to set
+        the `include_t1oo` attribute accordingly
+        """
+        parts = parse.urlparse(dsn)
+        params = parse.parse_qsl(parts.query, keep_blank_values=True)
+
+        include_t1oo_values = []
+        other_params = []
+        for key, value in params:
+            if key == "opensafely_include_t1oo":
+                include_t1oo_values.append(value)
+            else:
+                other_params.append((key, value))
```

Comment:
`parse_qs` would give us the list of `opensafely_include_t1oo` values without having to loop through `params`

```
In [1]: url = "mssql://foo/db?opensafely_include_t1oo=false&opensafely_include_t1oo=false&param1=1&param1=3&param2="
In [2]: parts = urlparse(url)
In [3]: params = parse_qs(parts.query, keep_blank_values=True)
In [4]: params
{'opensafely_include_t1oo': ['false', 'false'],
 'param1': ['1', '3'],
 'param2': ['']}
include_t1oo_values = params.pop("opensafely_include_t1oo", [])
```

Of course, then we'd still need to loop to get tuple pairs for `other_params`, so maybe it doesn't make much difference. It could be a one-liner, but I'm never sure nested list comprehensions are worth it for readability.  It's a simple one though, so :shrug: 
```
other_params =  [(k, val) for k, vals in params.items() for val in vals]
```

Reply:
I think you can actually pass the results from `parse_qs` directly into `urlencode` so we wouldn't even need to unpack. My silly reason for not doing that was that it wouldn't preserve parameter order for repeated params and I was obsessing over returning the URL as close to how it was supplied as possible.

But now I think about it that really _is_ a silly reason. I'll change it to your approach.